src/pushsource/_impl/backend/staged/staged_cloud.py

```diff
@@ -0,0 +1,141 @@
+
+        image_attrs = [
+            "dest",
+            "type",
+            "region",
+            "virtualization",
+            "volume",
+            "root_device",
+            "description",
+            "sriov_net_support",
+            "ena_support",
+            "uefi_support",
+            "public_image",
+            "release_notes",
+            "usage_instructions",
+            "recommended_instance_type",
+            "marketplace_entity_type",
+            "image_id",
+            "scanning_port",
+            "user_name",
+            "version_title",
+            "security_groups",
+            "access_endpoint_url",
+        ]
+
+        for key in image_attrs:
+            if key in resources:
+                image_kwargs[key] = resources.get(key)
+
+        return AmiPushItem(**image_kwargs)
+
+    def __build_azure_push_item(self, resources, src, origin, name):
+        build_resources = resources.get("build")
+        build_info = KojiBuildInfo(
+            name=build_resources.get("name"),
+            version=build_resources.get("version"),
+            release=build_resources.get("respin"),
+        )
+        image_kwargs = {
+            "name": name,
+            "src": src,
+            "description": resources.get("description"),
+            "build_info": build_info,
+            "origin": origin,
+        }
+        return VHDPushItem(**image_kwargs)
+
+    def __process_builds(self, build_dirs):
+        out = []
+        for bd in build_dirs:
+            yaml_path = os.path.join(bd, "resources.yaml")
+            try:
+                with open(yaml_path, "rt") as fh:
+                    raw = yaml.safe_load(fh)
+            except FileNotFoundError:
+                LOG.warning("No resources.yaml file found at %s (ignored)", yaml_path)
+                continue
+            if not raw:
+                continue
+            image_type = raw.get("type", "")
+            images_info = raw.get("images", [])
+
+            for image in images_info:
+                if "/" in image.get("path"):
+                    LOG.warning("Unexpected '/' in %s (ignored)", image_relative_path)
+                    continue
+                image_relative_path = os.path.join(bd, image.get("path"))
+                if image_type == "AMI":
+                    out.append(
+                        self.__build_ami_push_item(
+                            raw, image_relative_path, bd, image.get("path")
+                        )
+                    )
+                elif image_type == "VHD":
+                    out.append(
+                        self.__build_azure_push_item(
+                            raw, image_relative_path, bd, image.get("path")
+                        )
+                    )
+        return out
+
+    @handles_type("CLOUD_IMAGES")
+    def __cloud_push_item(self, leafdir, _, __):
+        builds = [f.path for f in os.scandir(leafdir.path) if f.is_dir()]
```

Comment:
Regarding the "have to provide an empty staged.yaml" issue, here's the problem I think you're having.

The implementation of `@handles_type` expects that every type handler is designed to deal with *files* only. Until your change, that was true everywhere.

Handlers are invoked like this:

```
    def __mixin_push_items(self, leafdir, metadata, delegate):
        out = []

        LOG.debug("Looking for files in %s", leafdir)

        for entry in scandir(leafdir.path):
            if entry.is_file():
                item = delegate(leafdir, metadata, entry)
                if item:
                    out.append(item)

        return out
```

``leafdir`` is ``<root>/destination/CLOUD_IMAGES``.  Also note that `scandir` isn't recursive.  So, it's designed to call the handler once for every file inside the leafdir (non-recursive).  In your case, there are zero files within the leafdir - only directories - so it doesn't do anything.  You can force it to do something by adding an empty `staged.yaml` - or adding *any* file in the directory for that matter (and one wonders what it's going to do if the user adds more than one file).

Making the user add a file into the directory just to satisfy this code isn't acceptable, the code should rather be modified to better suit your scenario.

I have two suggested approaches how to fix this:

### Suggestion 1

Move the responsibility for the "is_file" checks *inside* of the handlers.

So, the code invoking the handlers becomes like this:

```
        for entry in scandir(leafdir.path):
             item = delegate(leafdir, metadata, entry)
             if item:
                 out.append(item)
```

All existing handlers internally implement "if entry is not a file, do nothing" like this:

```
    @handles_type("COMPS")
    def __push_item(self, leafdir, _, entry):
        if entry.is_file():
          return CompsXmlPushItem(
              name=entry.name, src=entry.path, origin=leafdir.topdir, dest=[leafdir.dest]
          )
```

And your own handler implements "if entry is not a dir containing resources.yaml, do nothing" like this:

```
    @handles_type("CLOUD_IMAGES")
    def __cloud_push_item(self, leafdir, _, entry):
        if not entry.is_dir() or not os.path.exists(os.path.join(entry.path, "resources.yaml"):
            return
        # ...OK, we know `entry` is a dir containing a resources.yaml, process it
```

### Suggestion 2

Update the `@handles_type` decorator to take a parameter controlling the check for whether or not a dir entry should be processed.

Your own handler then might look something like this:

```
    @handles_type("CLOUD_IMAGES",
        accepts=lambda entry: entry.is_dir() and os.path.exists(os.path.join(entry.path, "resources.yaml")))
    def __cloud_push_item(self, leafdir, _, entry):
        # When we get here, we already know `entry` is a dir containing a resources.yaml
```

The default value of `accepts` could be set to `lambda entry: entry.is_file()` so that all existing handlers would continue to work how they do today, without having to touch their code.